ghproxy/ghcache/ghcache.go

```diff
@@ -214,6 +219,10 @@ func (tr *tokensRegistry) getRequestWaitDuration(tokenBudgetName string, getReq
 	if t, exists := tr.tokens[tokenBudgetName]; exists {
 		toQueue, duration = tr.calculateRequestWaitDuration(t, toQueue, getReq)
 	}
+	// ignore calculations, do not update map, useful when GH API has degraded performance
+	if duration > tr.backoffTime {
```

Comment:
> Please always add tests

The whole `ghcache` is written without any tests, I was the first one to include UTs there to test the algorithm. Ok, I will introduce them, but it's not like the same is happening thought prow..

> I am not sure I understand why just disabling throttling when you hit a treshold would be a good idea. Also:

Why not? I think GH API not working is an exceptional situation, and for 3 weeks, we had problems only with that.

What do you mean by docs? Where to write it?

Reply:
Adding to this disabling thing. If you do not have somebody (gh API) to handle your requests, the queue only grows. GH Api incident is resulting in something like this  (and the waiting times are only growing...):
![throttling](https://user-images.githubusercontent.com/78746728/159941102-f4992010-fdc4-4725-b3e4-f1d923de1203.png)

Reply:
> What I would expect a ratelimiter with a treshold to do when it hits that threshold is to drop the request, not to stop ratelimiting. That @petr-muller disabled the ratelimiter altogether seemed like a hotfix for the fact that we hadn't accounted for the situation of the gh api being down, not necesariliy like a good long-term solution that needs to be codified. So please, at minimum document somewhere why you took this approach over dropping requests.

Well, we could do that but we can just fall back to the limit concurrency thing that was before. I do not think it's a bad approach (to go level down with throttling). Although, yes that could be a solution. Waiting on more comments if we really want that. If not, I'll do better with commenting on that.

Reply:
I think this safety valve is necessary, not really sure whether I personally prefer dropping requests (which is the easy, non-suprising, explicit way) or what this PR is doing, which may result in some chaotic behavior, potentially triggering rate limiting and some kind of thundering herd behavior, but also potentially letting more requests through successfully. I think I would slightly prefer the dropping though, because it makes the behavior more predictable. Would it make sense to drop with a 429 and add a header that would trigger client retry (simulate the rate limiting).?

I think one of the problems with this PR is that it will by design result in wrong ordering of the requests (there is a queue of previous requests, but the requests coming after the backoff time are proxied immediately). No matter what the implementation is, the behavior around the boundary will be messy, oscillating between going over the boundary (getting fired right away or dropped) and under the boundary (scheduled at the end of the queue) - not sure if there's a way to avoid that. Maybe, if we're *not* dropping, we could simply schedule the requests at the backoff time at latest (instead of firing them right away)? That way we would create small clusters of requests that would be sent to GH API closer together, but that could be fine: at worst they get rate limited or something, get eventually retried...

Reply:
No 1. can lead to potentially different problems as timestamps are calculated and saved, I can't clearly see how it will affect future calculations (it might, I also thought about this solution) 
No 2. Seems reasonable though. If no objections I can try to implement that.

Reply:
@jmguzik did you confuse options one and two in your description? Fwiw Retry after is a delay in seconds. both 1. and 2. SGTM, no strong preference between the two

Reply:
@alvaroaleman actually I think a change of plans. Option 2 (RetryAfter) would require me to compose a fake github response with needed things to get it handled. Not sure if the error code is the only thing I need in the header (probably nope). As I have no idea what is needed and what not for such a message to get handled properly, I would rework this PR to support Option 1: Make the backoff behave like a maximum delay instead of a disable.

Reply:
Reworked to have option 1.